.github/workflows/ci.yml

```diff
@@ -99,6 +99,103 @@ jobs:
           name: test-results
           path: 'up-cpp/build/test/results/*.xml'
 
+# NOTE: Run dynamic analysis in unit tests
+  valgrind:
+    name: Run up-cpp tests with Valgrind
+    runs-on: ubuntu-latest
+    needs: build
+
+    steps:
+      - name: Get build artifacts
+        uses: actions/download-artifact@v4
+        with:
+          name: build-artifacts
+          path: up-cpp/build
+
+      - name: Install Valgrind
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y valgrind
+
+      - name: Run Valgrind Memcheck
+        continue-on-error: true
+        run: |
+          cd up-cpp/build
+          chmod +x bin/*
+          echo "Running Memcheck..."
+          exclude_tests=$(cat valgrind_exclude_tests.txt | tr '\n' '|' | sed 's/|$//')
+          echo "Excluding tests: $exclude_tests"
+          set +e
+          valgrind --leak-check=full \
+                   --show-leak-kinds=all \
+                   --track-origins=yes \
+                   --trace-children=yes \
+                   --verbose \
+                   --log-file=valgrind-memcheck.log \
+                   --gen-suppressions=all \
+                   ctest -E "$exclude_tests"
+
+          exit_code=$?
+          set -e
+
+          if [ $exit_code -ne 0 ]; then
+            touch memcheck_failed_output.log
+            echo "Some tests failed. Capturing the logs..."
+            cat Testing/Temporary/LastTest.log | tee -a memcheck_failed_output.log
+          else
+            echo "No tests failed."
+            touch memcheck_failed_output.log
+          fi
+
+      - name: Run Valgrind ThreadCheck
+        continue-on-error: true
+        run: |
+          cd up-cpp/build
+          echo "Running ThreadCheck..."
+          valgrind --tool=drd \
+                   --trace-children=yes \
+                   --log-file=valgrind-threadcheck.log \
+                   ctest
+          echo "ThreadCheck log:"
+          cat valgrind-threadcheck.log
+        
+      - name: Run Valgrind Helgrind
+        continue-on-error: true
+        run: |
+          cd up-cpp/build
+          echo "Running Helgrind..."
+          valgrind --tool=helgrind \
+                   --trace-children=yes \
+                   --log-file=valgrind-helgrind.log \
+                   ctest
+          echo "Helgrind log:"
+          cat valgrind-helgrind.log
+
+      - name: Run Valgrind DHAT
+        continue-on-error: true
+        run: |
+          cd up-cpp/build
+          echo "Running DHAT (Data Address Translation)..."
+          valgrind --tool=dhat \
+                   --trace-children=yes \
+                   --log-file=valgrind-dhat.log \
+                   ctest
```

Comment:
Should all of the tests have the `-E` flag to exclude the `valgrind_exclude_tests.txt` list?

Reply:
We can create different files and -E for each kind of failure in valgrind. The current implementation is only for memcheck, but others should follow similar logic.

I will rename this to something like `valgrind_exclude_memcheck_test.txt`